lmcache/experimental/lookup_server/redis_server.py

```diff
@@ -28,31 +47,128 @@ def __init__(self, config: LMCacheEngineConfig):
 
         self.connection = redis.Redis(host=self.host,
                                       port=self.port,
+                                      socket_timeout=1,
                                       decode_responses=True)
         logger.info(f"Connected to Redis lookup server at {host}:{port}")
         #decode_responses=False)
 
-    def _get_indexing_key(self, key: CacheEngineKey):
+        self.queue: asyncio.PriorityQueue[tuple[_WorkPriority, _WorkItem]] = (
+            asyncio.PriorityQueue())
+        self.loop = asyncio.new_event_loop()
+        self.thread = threading.Thread(target=self._run_loop, daemon=True)
+        self.thread.start()
+        asyncio.run_coroutine_threadsafe(self._requester(), self.loop)
+
+    async def _get_batch(self) -> AsyncGenerator[_WorkItem, None]:
+        """
+        Batch work items as an async generator.
+
+        :yields: WorkItem's one by one as they are retrieved.
+        """
+        # wait for first item in batch
+        _, item = await self.queue.get()
+        yield item
+
+        if item.op == _Op.LOOKUP:
+            timeout = 0.0  # lookups should not wait
+        else:
+            timeout = BATCH_TIMEOUT
+
+        for _ in range(BATCH_SIZE - 1):
+            try:
+                _, item = await asyncio.wait_for(self.queue.get(),
+                                                 timeout=timeout)
+                yield item
+                if timeout > 0 and item.op == _Op.LOOKUP:
+                    timeout = 0  # lookups should not wait
+            except asyncio.TimeoutError:
+                break
+
+    def _add_to_pipeline(self, pipe: redis.client.Pipeline, item: _WorkItem):
+        """
+        Add a request to a redis pipeline (batch of requests).
+
+        :param redis.client.Pipeline pipe: The redis pipeline to add to.
+
+        :param _WorkItem item: The item defining the request we want to add.
+        """
+        match item.op:
+            case _Op.INSERT:
+                pipe.hset(self._get_indexing_key(item.key),
+                          self.distributed_url,
+                          self._get_indexing_metadata(item.key))
+            case _Op.DELETE:
+                pipe.hdel(self._get_indexing_key(item.key),
+                          self.distributed_url)
+            case _Op.LOOKUP:
+                pipe.hgetall(self._get_indexing_key(item.key))
+
+    async def _requester(self):
+        """
+        _requester batches requests to be sent to redis
+        """
+        while True:
+            items = []
+            with self.connection.pipeline() as pipe:
+                async for item in self._get_batch():
+                    self._add_to_pipeline(pipe, item)
+                    items.append(item)
+                logger.debug(f"Sending a batch of {len(items)} requests")
+                results = pipe.execute()
```

Comment:
Is this call guaranteed to not throw? As far as I know - it could. Better wrap this with a try.

Reply:
You're right. I pushed a fix.
For the rest of the comments I will add a configuration parameter (plan for tomorrow).